core/src/main/java/org/apache/calcite/sql/SqlFunction.java

```diff
@@ -310,7 +310,9 @@ argTypes, argNames, getFunctionType(), SqlSyntax.FUNCTION,
           // if we succeed, the arguments would be wrapped with CAST operator.
           if (function != null) {
             TypeCoercion typeCoercion = validator.getTypeCoercion();
-            if (typeCoercion.userDefinedFunctionCoercion(scope, call, function)) {
+            if ((function.category == SqlFunctionCategory.USER_DEFINED_FUNCTION
```

Comment:
I'm not really sure why these judgments are made. Maybe I can guess a little, but it may not be very accurate. Should JavaDoc be added?

Reply:
JavaDoc comments can only be added to methods and classes. I suspect you are asking for regular Java comments.
Also, I am not sure what you mean by "judgments". I didn't make any judgments, I only attempted to fix a bug.

This is where the bug is located, in the absence of these checks the program fails with a class cast exception.
The implementation of the `userDefinedFunctionCoercion` already has a JavaDoc comment indicating that it is designed for user-defined functions.
And indeed, if I restrict the function to be invoked only on user-defined functions, as the inserted "if" does, the crash goes away.
I don't really know what additional comments to write which would clarify what this change does. I think the code is simple enough and the names are clear enough to describe what is going on.

Reply:
Thanks for your answer, I checked the code and it is indeed